Dockerfile.alpine

```diff
@@ -0,0 +1,17 @@
+FROM python:3.12-alpine3.20
+
+RUN apk upgrade --update && apk add --no-cache \
+  bash \
+  git \
+  cargo \
+  linux-headers
+
+WORKDIR /app
+
+COPY requirements.txt .
```

Comment:
the main differences with the other docker file are
* it uses the committed `requirements.txt` to guarantee a stable build
* this step is before the `COPY src` to optimize the usage of the layers - dependencies change less frequently than the code

Reply:
in my view (but i'm a total noob at docker, so take it with a grain of salt) requirements.txt should be generated at image-build time, not committed.

the ideal way would be run tests -> if ok -> generate temporary requirements.txt -> build image

we don't have tests at the moment, but i think the above approach is valid. no ?

Reply:
Please keep in mind that i have very little experience with python, but this has nothing to do with docker in general. Is `pyproject.toml` a new alternative to  `requirements.txt`?

Either way they should be committed with the version of each lib in the source code, the reason is that if you build your project *today* all the dependencies you install/use will have the latest version available *today*. If you re-build it in 6 months they will likely be old and if someone will pull the project from scratch (or you'll kick a new build in the ci) the likelihood that it won't build and there will be breaking changes will only increase over time.
It's just a way to keep the source code consistent with the lib version, that's all.
If `pyproject.toml` allows to pin the version and it's enough to build the project without the `requirements.txt` file, then you don't need it.

Please let me know if it makes sense or there is a better alternative, thanks

Reply:
pyproject.toml is just an "enhanced version" of the "old" setup.py and includes the packages list (requirements.txt), and it of course allows to pin version. but...

> Either way they should be committed with the version of each lib in the source code, the reason is that if you build your project today all the dependencies you install/use will have the latest version available today. 

this is what i want, indeed. no point in use outdated stuff/libs unless they are no more mantained.

> . If you re-build it in 6 months they will likely be old and if someone will pull the project from scratch (or you'll kick a new build in the ci) the likelihood that it won't build and there will be breaking changes will only increase over time.
It's just a way to keep the source code consistent with the lib version, that's all.

this is true aswell, and this is where the CI tests would help greatly. 
unfortunately, we do not have them yet :(

this way, if the tests succeed, we would ensure to run with the latest versions of all the packages we use.
if the tests fail, we would notice and fix our code to be compatible (or, freeze the offending packages temporarly until we have a fix).

so, in the end, i prefer to keep pyproject.toml with unpinned dependencies, and pin only at build time (i know we currently miss a test suite, and we are aware of the implications....)

Reply:
thanks for the feedback, so probably all you need from this pr is `.github/workflows/docker-ci.yml` which by default looks for a `Dockerfile` in the root path and an extra step before to generate the requirements file

this is really up to you, if you don't wanna build a snapshot image on every master build you can then split the workflow to publish stable versions only on a new tag and run all the times only the build - i wrote and used this action in other projects and it does the job

At this stage I would probably leave this pr as an example cos you need to add the secrets and it's easier to test it for you

Once the image is published on docker hub please let me know and i will create a new pr to add the helm chart to deploy it on kube 😉

Reply:
basically, i would need a way to run update_requirements_txt.sh inside the env, whenever the build is triggered. this would freeze the requirements and create the image (this is what you do in your notes at the top comment). can you include this in the PR ?

Reply:
anyway, merged, we will handle the requirements thing later :) thanks for the PR !

Reply:
thanks, as you prefer! i was actually planning to look into it. Please let me know if you need any support, I'm pretty sure the build is going to fail
* probably this should be `develop` https://github.com/mentat-is/gulp/blob/develop/.github/workflows/docker-ci.yml#L6
* try to add `file: Dockefile.alpine` if you want to use this image https://github.com/mentat-is/gulp/blob/develop/.github/workflows/docker-ci.yml#L53, see https://github.com/docker/build-push-action?tab=readme-ov-file#customizing
* make this dynamic i.e. only on tag `${{ github.ref_type == "tag" }} https://github.com/mentat-is/gulp/blob/develop/.github/workflows/docker-ci.yml#L57 so that the build will run but you won't push snapshots

please let me know if you need any help